### PR TITLE
Thinning of electromagnetic processes

### PR DESCRIPTION
**Pull request by rafaelab**
Implementation of thinning. Supersedes #277.
DINT/EleCa removal will be done in another pull request.

I suggest we wait for synchrotron thinning before merging. But this may take some time.

### CODE REVIEW

**Comment by lukasmerten** [comment]
@rafaelab, this looks good from my side now. It would be nice if you add an example when the thinning is integrated into the main branch. 

When you implemented the thinning for synchrotron radition I will check that too.

**Comment by rafaelab** [comment]
Quick update:
Even with full thinning, synchrotron is still extremely slow and somewhat memory-consuming. I couldn't even finish testing the implementation within a reasonable time.
I'm working on a "second thinning" to speed things up.

**Comment by rafaelab** [comment]
The new commits introduce thinning for synchrotron.
The procedure is in part the same one used for the other EM processes. But this doesn't improve the situation a lot (only for a few steps), so another type of procedure was devised. 

The module computes dE/dx. The energy lost in the step of length dx is dE. Synchrotron photons are emitted until dE~0. The additional sampling consists in assuming that we can sample only a few of these photons with energy Ephoton at each step, such that they add to dE1<dE. The loop is stopped after a given number of samples has been reached. If this sample is large enough and the step small enough, we can add (after the loop) secondary photons of energy Ephoton with weight dE/dE1, to conserve the total energy loss at each step.

File: src/module/SynchrotronRadiation.cpp
Comment (lukasmerten) [inline src/module/SynchrotronRadiation.cpp:17]: What was the reason for changing the default value of the threshold?

File: src/module/SynchrotronRadiation.cpp
Comment (lukasmerten) [inline src/module/SynchrotronRadiation.cpp:26]: Same as above

File: src/module/SynchrotronRadiation.cpp
Comment (rafaelab) [reply src/module/SynchrotronRadiation.cpp:17]: This choice was arbitrary. The previous value had no physical meaning, neither does this. 
It would make sense to choose either 1 GeV, which is roughly the minimum energy that we treat with CRPropa, or 1 keV, since X-ray energies are typical for synchrotron emission.

File: src/module/SynchrotronRadiation.cpp
Comment (lukasmerten) [inline src/module/SynchrotronRadiation.cpp:188]: This weight `w1` account for that part of the energy loss `dE` that was *left over* because of the limited number of photons that have been drawn, due to `maximumSamples`, right?

File: src/module/SynchrotronRadiation.cpp
Comment (lukasmerten) [inline src/module/SynchrotronRadiation.cpp:194]: The last part of the weight `1/pow(f, thinning)` then accounts for the fact that only a fraction of the sampled photons is accepted, as in the modules with thinning, right?

File: src/module/SynchrotronRadiation.cpp
Comment (lukasmerten) [reply src/module/SynchrotronRadiation.cpp:17]: In that case I would prefer to keep the old value. Just to prevent any possible confusion.

File: src/module/SynchrotronRadiation.cpp
Comment (rafaelab) [reply src/module/SynchrotronRadiation.cpp:188]: Exactly. It is just a correction factor to ensure energy conservation based on a sample of all the photons that would have been drawn.

The energy to be lost is dE0. But with the sampling only dE was lost within this step. The closer dE is to dE0, the lower the weights required, since the sample is large enough. For a tiny sample, only a few particles are chosen and dE/dE0 is very small, thus requiring larger weights.

File: src/module/SynchrotronRadiation.cpp
Comment (rafaelab) [reply src/module/SynchrotronRadiation.cpp:194]: Yes. This is essentially the usual thinning method used in the other EM modules. It doesn't help a lot in most cases, but combined with the sampling, it increases the performance significantly.

**Comment by rafaelab** [comment]
Thanks, @lukasmerten.
But I'd like to finish running the very long tests for the synchrotron before it is merged, so please wait a couple more weeks. So far all the scenarios I checked look okay, but you never know...
I will also change the default value back to 1e7 eV, although I would be in favour of setting it to 1e9 eV since we rarely go below that in CRPropa anyways.

**Comment by lukasmerten** [comment]
@rafaelab Doing all the tests before merging is good. Just let me know, when I can help with something.

File: test/testInteraction.cpp
Comment (lukasmerten) [inline test/testInteraction.cpp:574]: Does it make sense to use here and in the other test the same value for the currentStep? This of course should then be the smallest one: 1e3Mpc, which is probably large enough for all simulation setups anyway, isn't it?

Comment (rafaelab) [reply test/testInteraction.cpp:574]: For pair production it doesn't really matter which step is used and the tests converge fast.
For inverse Compton scattering and triplet pair production, if the step is very large it takes ages to finilise the test. A suitable value is ~1e4 Mpc (Hubble horizon).

Is your suggestion to change this step to the same value for all EM processes?

Comment (lukasmerten) [reply test/testInteraction.cpp:574]: Yes, that was my suggestion.

**Review by lukasmerten — APPROVED** [review]

**Comment by lukasmerten** [comment]
When @rafaelab has finished his testing, I think we can merge this pull request, after the conflicts have been solved.

**Comment by rafaelab** [comment]
Ah!  I didn't realise there were conflicts now. There weren't before. I will fix them soon.

**Comment by lukasmerten** [comment]
@rafaelab Could you also update the changelog? Sorry, for not mentioning it earlier.

**Comment by rafaelab** [comment]
> @rafaelab Could you also update the changelog? Sorry, for not mentioning it earlier.

Thanks for the reminder, @lukasmerten.
From my side this can now be merged. 
For the record, in the future my implementation should be compared with #304

**Pull request merged**
